src/repr/src/row/encoding.rs

```diff
@@ -72,73 +44,335 @@ impl Codec for Row {
+                self.push(Datum::from(n))
+            }
+            None => return Err("unknown datum type".into()),
+        };
+        Ok(())
+    }
+}
 
-        // NB: The read calls modify buf to truncate off what they read, so
-        // index 0 now corresponds to the part of the original buf immediately
-        // after the encoded len.
-        let row_data = buf
-            .get(0..len)
-            .ok_or_else(|| format!("wanted {} row data bytes but had {}", len, buf.len()))?;
+impl TryFrom<&ProtoRow> for Row {
+    type Error = String;
 
-        // SAFETY: This was serialized with Row::encode at the same version.
-        let row = unsafe { Row::from_bytes_unchecked(row_data.to_owned()) };
+    fn try_from(x: &ProtoRow) -> Result<Self, Self::Error> {
+        // TODO: Try to pre-size this.
+        let mut row = Row::default();
+        for d in x.datums.iter() {
+            row.try_push_proto(d)?;
+        }
         Ok(row)
     }
 }
 
 #[cfg(test)]
 mod tests {
+    use chrono::{DateTime, NaiveDate, NaiveTime, Utc};
     use persist_types::Codec;
+    use uuid::Uuid;
 
+    use crate::adt::array::ArrayDimension;
+    use crate::adt::interval::Interval;
+    use crate::adt::numeric::Numeric;
     use crate::{Datum, Row};
 
     // TODO: datadriven golden tests for various interesting Datums and Rows to
     // catch any changes in the encoding.
 
     #[test]
-    fn decode_errors() {
-        let row = Row::pack(vec![Datum::Int64(7)]);
+    fn roundtrip() {
+        let mut row = Row::pack(vec![
+            Datum::False,
+            Datum::True,
+            Datum::Int16(1),
+            Datum::Int32(2),
+            Datum::Int64(3),
+            Datum::Float32(4f32.into()),
+            Datum::Float64(5f64.into()),
+            Datum::Date(NaiveDate::from_ymd(6, 7, 8)),
+            Datum::Time(NaiveTime::from_hms(9, 10, 11)),
+            Datum::Timestamp(
+                NaiveDate::from_ymd(12, 13 % 12, 14).and_time(NaiveTime::from_hms(15, 16, 17)),
+            ),
+            Datum::TimestampTz(DateTime::from_utc(
+                NaiveDate::from_ymd(18, 19 % 12, 20).and_time(NaiveTime::from_hms(21, 22, 23)),
+                Utc,
+            )),
+            Datum::Interval(Interval {
+                months: 24,
+                duration: 25,
+            }),
+            Datum::Bytes(&[26, 27]),
+            Datum::String("28".into()),
+            Datum::from(Numeric::from(29)),
+            Datum::JsonNull,
+            Datum::Uuid(Uuid::from_u128(30)),
+            Datum::Dummy,
+            Datum::Null,
+        ]);
+        row.push_array(
+            &[ArrayDimension {
+                lower_bound: 2,
+                length: 2,
+            }],
+            vec![Datum::Int32(31), Datum::Int32(32)],
+        )
+        .expect("valid array");
+        row.push_dict_with(|row| {
+            row.push(Datum::String("33"));
+            row.push(Datum::Int32(34));
```

Comment:
can you add a bunch more keys and values here to try to tickle the hashmap bug?

Reply:
Will do. Verified the new version catches the bug